File: 模板/python/UnionFind.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.part = n

    def find(self, i):
        if self.parent[i] != i:
            self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i, j):
        x, y = self.find(i), self.find(j)
        if x != y:
            self.size[y] += self.size[x]
            self.parent[x] = self.parent[y]
            self.part -= 1

File: 模板/cpp/UnionFind.cpp
struct UnionFind {
    vi parent, size;
	int part;
	UnionFind (int n) : part(n) {
        parent = vi(n), size = vi(n);
		rep(i, 0, n) parent[i] = i, size[i] = 1;
	};
	~UnionFind() {}

	int find(int i) {
		if (parent[i] != i) {
			parent[i] = find(parent[i]);
		}
		return parent[i];
	}

	void _union(int i, int j) {
		int x = find(i), y = find(j);
		if (x != y) {
			size[y] += size[x];
			parent[x] = parent[y];
			part -= 1;
		}
	}
};
